performance/admin.py

```diff
@@ -1,9 +1,168 @@
+                q2_dpme_coordinator_comments='',
+                q2_treasury_comments='',
+
+                q3_target=indicator_data[23],
+                q3_actual_output=indicator_data[24],
+                q3_deviation_reason=indicator_data[25],
+                q3_corrective_action=indicator_data[26],
+                q3_national_comments=indicator_data[28],
+                q3_otp_comments=indicator_data[27],
+                q3_dpme_coordinator_comments='',
+                q3_treasury_comments='',
+
+                q4_target=indicator_data[29],
+                q4_actual_output=indicator_data[30],
+                q4_deviation_reason=indicator_data[31],
+                q4_corrective_action=indicator_data[32],
+                q4_national_comments=indicator_data[34],
+                q4_otp_comments=indicator_data[33],
+                q4_dpme_coordinator_comments='',
+                q4_treasury_comments='',
+
+                annual_target=indicator_data[35],
+                annual_aggregate_output='',
+                annual_pre_audit_output=indicator_data[37],
+                annual_deviation_reason=indicator_data[38],
+                annual_corrective_action=indicator_data[39],
+                annual_otp_comments=indicator_data[40],
+                annual_national_comments=indicator_data[41],
+                annual_dpme_coordincator_comments='',
+                annual_treasury_comments='',
+                annual_audited_output=indicator_data[42],
+
+                sector=indicator_data[0],
+                programme_name=indicator_data[2],
+                subprogramme_name=indicator_data[3],
+                frequency=indicator_data[5],
+                type=indicator_data[7],
+                subtype=indicator_data[8],
+                mtsf_outcome=indicator_data[9],
+                cluster=indicator_data[10],
+                uid=indicator_data[43],
+            )
+            num_imported = num_imported + 1
+        elif department_name not in not_matching_departments:
+            not_matching_departments.append(department_name)
+
+    obj_to_update = models.EQPRSFileUpload.objects.get(id=obj_id)
+
+    obj_to_update.num_imported = num_imported
+    obj_to_update.num_not_imported = total_record_count - num_imported
+    obj_to_update.import_report = generate_import_report(True, None, not_matching_departments)
+    obj_to_update.save()
+
+
+def validate_report_type(full_text, obj_id):
+    validated = False
+    for report_type in VALID_REPORT_TYPES:
+        validated = validated or (report_type in full_text)
+
+    if not validated:
+        obj_to_update = models.EQPRSFileUpload.objects.get(id=obj_id)
+        obj_to_update.num_imported = None
+        obj_to_update.num_not_imported = None
+        obj_to_update.import_report = generate_import_report(False, None, [])
+        obj_to_update.save()
+
+    return validated
+
+
+def validate_frictionless(data, obj_id):
+    report = validate(data)
```

Comment:
I think we should drop frictionless for the first iteration - especially if we switch to using key-based field lookup instead of index-based.

What I wanted to validate using frictionless was the presence of the required fields - see the suggestions below. But doing that in-memory on a server that might be doing this concurrently isn't obvious from the docs. I think instead of to_yaml in the docs links below can be swapped with to_dict. But I think we can waste a lot of time on this and I think we'll be safe enough with key-based lookup and key errors for a first iteration, and while we get the PR ready to ship, we can start the next iteration that will provide better user feedback using frictionless while we figure out the API.

For information:

I used the following schema to validate this excel file:

```json
{
  "fields": [
    {
      "type": "string",
      "name": "government",
      "constraints": {
        "pattern": "^([a-zA-Z-]+ )*[a-zA-Z-]+$"
      }
    },
    {
      "type": "string",
      "name": "group_id",
      "constraints": {
        "pattern": "^[a-z-]+$"
      }
    },
    {
      "type": "string",
      "name": "dataset_name"
    },
    {
      "type": "string",
      "name": "department_name"
    },
    {
      "type": "string",
      "name": "dataset_title"
    },
    {
      "type": "string",
      "name": "resource_name"
    },
    {
      "type": "string",
      "name": "resource_format"
    },
    {
      "type": "string",
      "name": "resource_url"
    }
  ]
}
```

[epre-budget-vote-docs-bulk-2022-06-22-upload.xlsx](https://github.com/vulekamali/datamanager/files/10337281/epre-budget-vote-docs-bulk-2022-06-22-upload.xlsx)
 with the command 

```bash
poetry run frictionless validate --schema-sync --schema budget-actual/bulk-budget-doc-metadata.json --limit-errors 3 budget-actual/epre-budget-vote-docs-bulk-2022-06-22-upload.xlsx
```

and I got the following output:

```
# -------
# invalid: budget-actual/epre-budget-vote-docs-bulk-2022-06-22-upload.xlsx 
# -------

## Summary 

The document was partially validated because of one of the limits
* limit errors 
* memory Limit

+-------------------------------------+-----------------------------------------------------------------+
| Description                         | Size/Name/Count                                                 |
+=====================================+=================================================================+
| File name (Not Found)               | budget-actual/epre-budget-vote-docs-bulk-2022-06-22-upload.xlsx |
+-------------------------------------+-----------------------------------------------------------------+
| File size                           | N/A                                                             |
+-------------------------------------+-----------------------------------------------------------------+
| Total Time Taken (sec)              | 0.136                                                           |
+-------------------------------------+-----------------------------------------------------------------+
| Rows Checked(Partial)**             | 72                                                              |
+-------------------------------------+-----------------------------------------------------------------+
| Total Errors                        | 3                                                               |
+-------------------------------------+-----------------------------------------------------------------+
| Constraint Error (constraint-error) | 3                                                               |
+-------------------------------------+-----------------------------------------------------------------+

## Errors 

+-------+---------+------------+----------------------------------------------------+
|   row |   field | code       | message                                            |
+=======+=========+============+====================================================+
|    70 |       1 | constraint | The cell "Mpumalanga " in row at position "70" and |
|       |         | -error     | field "government" at position "1" does not        |
|       |         |            | conform to a constraint: constraint "pattern" is   |
|       |         |            | "^([a-zA-Z-]+ )*[a-zA-Z-]+$"                       |
+-------+---------+------------+----------------------------------------------------+
|    71 |       2 | constraint | The cell "budget-vote_documents" in row at         |
|       |         | -error     | position "71" and field "group_id" at position "2" |
|       |         |            | does not conform to a constraint: constraint       |
|       |         |            | "pattern" is "^[a-z-]+$"                           |
+-------+---------+------------+----------------------------------------------------+
|    72 |       1 | constraint | The cell "Mpumalanga " in row at position "72" and |
|       |         | -error     | field "government" at position "1" does not        |
|       |         |            | conform to a constraint: constraint "pattern" is   |
|       |         |            | "^([a-zA-Z-]+ )*[a-zA-Z-]+$"                       |
+-------+---------+------------+----------------------------------------------------+
```

**only include the `name` fields** - the `type` and `constraint` fields are optional and we can skip them for now. 

The `--schema-sync` argument means "order the schema fields as per the data" so that column order doesn't break validation. I thin in the python method the argument is `validate(..., sync_schema=True)` https://github.com/frictionlessdata/framework/issues/380#issuecomment-698954928

https://v4.framework.frictionlessdata.io/docs/guides/validation-guide#validation-report

Reply:
removed validation